src/analysis/scoring.cc

```diff
@@ -0,0 +1,305 @@
+#include "analysis/scoring.h"
+#include <algorithm>
+#include <vector>
+
+#include "analysis/analysis.h"
+#include "analysis/util.h"
+#include "analysis/yakus.h"
+#include "types/gamestate.h"
+#include "types/pieces.h"
+#include "types/piecetype.h"
+#include "types/score.h"
+
+namespace mahjong {
+
+namespace {
+const int kLowHanBasicPoints = 2000;
+const int kManganBasicPoints = 2000;
+const int kManganMinHan = 5;
+const int kHanemanBasicPoints = 3000;
+const int kHanemanMinHan = 6;
+const int kBaimanBasicPoints = 4000;
+const int kBaimanMinHan = 8;
+const int kSanbaimanBasicPoints = 6000;
+const int kSanbaimanMinHan = 11;
+const int kYakumanBasicPoints = 8000;
+
+const int kPinfuDiscard = 30;
+const int kPinfuSelfdraw = 20;
+const int kConcealedDiscard = 25;
+const int kOpenOrSelfdraw = 20;
+
+const int kSimplepon = 2;
+const int kTermHonorpon = 4;
+const int kSimplekan = 8;
+const int kTermHonorkan = 16;
+
+const int kConcealedMul = 2;
+const int kCsimplepon = kSimplepon * kConcealedMul;
+const int kCtermHonorpon = kTermHonorpon * kConcealedMul;
+const int kCsimplekan = kSimplekan * kConcealedMul;
+const int kCtermHonorkan = kTermHonorkan * kConcealedMul;
+
+const int kDragonseatprevalentwind = 2;
+const int kEdgeclosedpairwait = 2;
+const int kSelfdraw = 2;
+const int kOpenpinfu = 2;
+
+const int kLowedgewait = 3;
+const int kHighedgewait = 7;
+
+const int kFuRounding = 10;
+
+const std::vector<mahjong::yakuFunc> kYakuFunctions = {
+    isFullyConcealedHand,
+    isPinfu,
+    isPureDoubleChi,
+    isAllSimples,
+    isMixedTripleChi,
+    isPureStraight,
+    isPrevalentWind,
+    isSeatWind,
+    isGreenDragon,
+    isWhiteDragon,
+    isRedDragon,
+    isOutsideHand,
+    isAfterAKan,
+    isRobbingAKan,
+    isBottomOfTheSea,
+    isSevenPairs,
+    isTriplePon,
+    isThreeConcealedPons,
+    isThreeKans,
+    isAllPons,
+    isHalfFlush,
+    isLittleThreeDragons,
+    isAllTerminalsAndHonors,
+    isTerminalsInAllSets,
+    isTwicePureDoubleChi,
+    isBlessingOfMan,
+    isFullFlush,
+};
+const std::vector<yakuFunc> kYakumanFunctions = {
+    isThirteenOrphans, isNineGates,        isBlessingOfHeaven,
+    isBlessingOfEarth, isFourConcealedPon, isFourKans,
+    isAllGreen,        isAllTerminals,     isAllHonors,
+    isBigThreeDragons, isLittleFourWinds,  isBigFourWinds,
+    isMaxBranches,
+};
+
+}  // namespace
+
+Score scoreHand(const GameState& state, int player) {
```

Comment:
A lot of this is reorganization so maybe out of scope for this pr, but can we get some scoring unit tests to verify everything integrates together correctly?